speechbrain/core.py

```diff
@@ -1024,60 +1093,92 @@ def fit_batch(self, batch):
-            if self.bfloat16_mix_prec:
+        amp = AMPConfig.from_name(self.precision)
+        should_step = (self.step % self.grad_accumulation_factor) == 0
+
+        with self.no_sync(not should_step):
+            if self.use_amp:
                 with torch.autocast(
-                    device_type=torch.device(self.device).type,
-                    dtype=torch.bfloat16,
+                    dtype=amp.dtype, device_type=torch.device(self.device).type,
                 ):
-                    outputs = self.compute_forward(batch, Stage.TRAIN)
-                    loss = self.compute_objectives(outputs, batch, Stage.TRAIN)
+                    outputs = self.compute_forward(batch, sb.Stage.TRAIN)
             else:
-                outputs = self.compute_forward(batch, Stage.TRAIN)
-                loss = self.compute_objectives(outputs, batch, Stage.TRAIN)
-
-            if self.check_gradients(loss):
-                valid_loss = True
-                self.valid_step += 1
-
-            should_step = self.valid_step % self.grad_accumulation_factor == 0
-            if valid_loss:
-                with self.no_sync(not should_step):
-                    (loss / self.grad_accumulation_factor).backward()
-                if should_step:
-                    self.optimizer.step()
-                    self.zero_grad()
-                    self.optimizer_step += 1
+                outputs = self.compute_forward(batch, sb.Stage.TRAIN)
+
+            # The objectives are removed from Autocast to avoid
+            # potential numerical instabilities with CTC loss, etc.
+            loss = self.compute_objectives(outputs, batch, sb.Stage.TRAIN)
+
+            scaled_loss = self.scaler.scale(
+                loss / self.grad_accumulation_factor
+            )
+            scaled_loss.backward()
+
+        if should_step:
+            self.optimizers_step()
 
         self.on_fit_batch_end(batch, outputs, loss, should_step)
         return loss.detach().cpu()
 
+    def check_gradients(self):
+        """ Checks if the gradients are finite. If not, it will emit a warning and set them to zero."""
+        for param in self.modules.parameters():
+            if param.requires_grad and param.grad is not None:
+                if not torch.isfinite(param.grad).all():
+                    param.grad = None
+                    logger.warning(
+                        f"Gradients {param.name} contain NaN or Inf. Setting to None."
+                    )
+
+    def freeze_optimizers(self, optimizers):
+        """By default, this method returns the passed optimizers.
+        Override this method if you want to freeze some optimizers
+        during training. To do so, return a of active optimizers.
+        """
+        return optimizers
+
+    def on_optimizers_step_end(self):
+        """Gets called at the end of ``optimizers_step()``.
+        Override this method to do something at the end of ``optimizers_step()``.
+        """
+        pass
```

Comment:
Are there cases that `on_optimizers_step_end()` solves that `on_fit_batch_end()` does not? For learning rate updates etc, the latter should work just fine, no?

Reply:
Good point. What surprises me the most is that no one actually is using `on_fit_batch_end()`. Thanks for raising this point. I will use the function and delete `on_optimizers_step_end()`. 

One question though: do you think we should move https://github.com/speechbrain/speechbrain/pull/2010/files#diff-d2ad56ee927e543eceb0f9528eea488134956a963c60d845d72a92d69ae0187aR1118 inside `on_fit_batch_end()`? I guess it make sense from a semantic point of view. We are using opt steps at the end of the fit_batch.

Reply:
The `on_fit_batch_end` was added some time after the first release, so possibly at the time that some overrides were created, the function was not available yet.

Reply:
If we put the call to `optimizer_step` inside `on_fit_batch_end` then I suppose people could override to add code right before it's called as well as right after. I'm not really sure how often that would be needed.

Reply:
The nice part about having small (or empty) functions to be overridden is that it reduces code copying

Reply:
Hmm Yes. Actually what I'm thinking about is that, in some cases people are using LR annealing. Which means, they have to override `on_fit_batch_end`. If we move `optimizer_step` inside `on_fit_batch_end` then it means they will have to add `optimizer_step` as well. Maybe it's better to leave it in `fit_batch` and use `on_fit_batch_end` only for specific cases like LR annealing. What do you think?

Reply:
BTW, side note, I changed the name of the flag to be `--precision` instead of `--autocast`. I think this is more intuitive to everyone, furthermore, autocast for many people, including me, means fp16/bf16 not fp32. PyTorch Lightning is also using `--precision`.